chore: parallelize follow-up email sending with Promise.all

0xajinkya opened the pull request:
## Type of change

<!--- What types of changes does your code introduce? Put an `x` in the box that apply: -->

- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] 🔨 Breaking change
- [ ] 🐛 Bug fix
- [ ] 🌟 New feature

## Description

<!--- Describe your changes in detail -->
Previously, follow-up emails were sent sequentially inside a for...of loop, leading to slower execution. This change uses Promise.all to send emails in parallel, significantly improving the speed of the process.

This PR improves the performance of the sendFollowUpEmails function by replacing the sequential email sending (using a for...of loop) with parallel sending using Promise.all.
This change speeds up the process, especially when there are many users to email, without affecting the functionality.

### A picture tells a thousand words (if any)

### Before this PR

![Screenshot from 2025-04-28 03-40-11](https://github.com/user-attachments/assets/7c9ad5e5-0c98-4eba-971d-d0f07fa704de)

### After this PR

![Screenshot from 2025-04-28 03-40-40](https://github.com/user-attachments/assets/4d09360b-16d7-4317-b229-deb937f6aa1e)

### Related Issue (optional)

<!--- Please link to the issue here: -->

Commit:
chore: parallelize follow-up email sending with Promise.all

Previously, follow-up emails were sent sequentially inside a for...of loop, leading to slower execution. This change uses Promise.all to send emails in parallel, significantly improving the speed of the process.

thibaultleouay commented:
Thanks for you PR 

My concern if we parallelize every send email we will hit resend rate limit 
what we could do but it might require a bit more work is to use sendBatch 
https://resend.com/docs/api-reference/emails/send-batch-emails

0xajinkya commented:
> Thanks for you PR
> 
> My concern if we parallelize every send email we will hit resend rate limit what we could do but it might require a bit more work is to use sendBatch https://resend.com/docs/api-reference/emails/send-batch-emails

Hey @thibaultleouay, I am looking into this. Will keep you updated.

Commit:
feat(emails): implement batch follow-up emails with rate limiting

Improve the follow-up email system by:
- Optimizing database queries to select only necessary email fields
- Implementing batched email sending (80 emails per batch)
- Adding proper rate limit detection and handling
- Improving error logging and recovery

0xajinkya commented:
Hey @thibaultleouay,

Can you look here?

Following are the key changes:
1. Database optimization: Now only selecting the email field from users instead of fetching complete user records, reducing memory usage and database load.

2. Batch processing: Implemented batching of emails (80 per batch) to prevent overwhelming the email service and improve throughput.

3. Rate limit handling: Added specific detection of rate_limit_exceeded errors, which stops further sending attempts when limits are hit to prevent wasted API calls.

4. Error handling improvements: Enhanced error catching with more specific error types and better logging of failures.

5. Process visibility: Added additional logging to track batch progress.

These changes should make the follow-up email process more reliable and prevent issues when sending to large numbers of users. The batching approach also provides better visibility into the sending process and enables easier recovery when errors occur.

thibaultleouay commented on apps/workflows/src/cron/emails.ts:
```diff
@@ -14,16 +14,30 @@ export async function sendFollowUpEmails() {
+  // Chunk emails into batches of 100
+  const batchSize = 80;
+  for (let i = 0; i < validEmails.length; i += batchSize) {
+    const batch = validEmails.slice(i, i + batchSize) as string[];
```
not a huge fan of `as string[]`, cant we auto infer it instead?

0xajinkya replied:
Latest change takes care of this. Please have a look.

thibaultleouay commented on packages/emails/src/client.tsx:
```diff
@@ -40,8 +40,43 @@ export class EmailClient {
+          throw result.error;
+        }
+
+        console.error(`Batch send error:`, result.error);
```
I would also log the error above  move this after line 59

0xajinkya replied:
The thing is, if the batch send fails specifically due to **rate limit**, it is intuitional that subsequent batches will fail as well. So, that's why I am not logging that particular error. And, instead handling it in the loop itself (Logging it and breaking from the loop). Any other error, are being simply consolled.

thibaultleouay commented on packages/emails/src/client.tsx:
```diff
@@ -40,8 +40,43 @@ export class EmailClient {
+      }
+
+      console.log(`Sent follow up email to ${req.to}`);
+    } catch (err: any) {
```
not sure about the try/catch

0xajinkya replied:
This is explained in the latest reply to above conversation. Have a look.

Commit:
fix: improve follow-up email sending reliability and error handling

The changes improve email delivery reliability by:

- Enhancing email validation with stronger type guards
- Adding proper break statement for rate limit errors to prevent wasteful API calls
- Simplifying error flow control with early returns and consistent error handling
- Restructuring try/catch blocks for better exception management
- Improving TypeScript type safety with more specific error handling

These modifications ensure the system gracefully handles rate limiting while providing clearer logs for debugging. Email validation now properly checks for empty strings, preventing attempts to send to invalid addresses.

0xajinkya commented:
Any updates @thibaultleouay?

thibaultleouay committed:
minor change

thibaultleouay commented:
@0xajinkya  Just pushed some update in your pr let me know if you have any question

Commit:
fix(followup-emails): added type checking for email as string

Initially, we were not checking type of email and calling trim() on it, this improvision check whether email is really a string and then calls trim() on it.

0xajinkya commented:
Hey @thibaultleouay, All good, added a minor fix. Also, I am wondering why are we doing this:

` const validEmails = users
    .map((u) => u.email)
    .filter((email) => email !== null)
    //  I don't know why but I can't have both filter at the same time
    .filter((email) => typeof email === "string" && email.trim() !== "");
`

This seems repetitive.

thibaultleouay commented:
@0xajinkya  We are filtering null first, then removing empty email

thibaultleouay commented on apps/workflows/src/cron/emails.ts:
```diff
@@ -14,16 +14,33 @@ export async function sendFollowUpEmails() {
+    .map((u) => u.email)
+    .filter((email) => email !== null)
+    //  I don't know why but I can't have both filter at the same time
+    .filter((email) => typeof email === "string" && email.trim() !== "");
```
![CleanShot 2025-05-01 at 12 01 49@2x](https://github.com/user-attachments/assets/77c89dbc-b5fd-4a64-b145-1c25950e18ec)

I dont understand the typeof email here, because the type is already infered ?

0xajinkya replied:
Yes, it was totally not needed. It was error from my end. Changing it.

Commit:
minor change: removed un-necassary type checking for email

autofix-ci[bot] committed:
ci: apply automated fixes

thibaultleouay commented:
Thanks for your pr :)

0xajinkya commented:
It's a pleasure, looking forward more such contributions.
